### PR TITLE
Set data type

### PR DESCRIPTION
**Pull request by edualonso**
This PR adds support for the new `RealmSet` data type.

- Sets support all Realm primitive types supported internally by `Mixed` - see https://github.com/realm/realm-java/pull/7398.
- Sets cannot contain repeated values.
- Values can be `null`, except in the case of `RealmModel`.
- Embedded objects are not supported.
- Sets with the `@Required` annotation will work just like `RealmList`s and will enforce a non-nullability constraint on the elements it contains. The `@Required` annotation is not allowed if the dictionary contains `RealmModel` or `Mixed` elements.
- `RealmModel`s containing `RealmSet` fields cannot be added to Realm using any of Realm's `insertX` methods nor can they be created from JSON sources using any of the `createFromJSON` yet.

Internally, the architectural implementation goes as follows:
```
                      RealmSet<E>
                           |
                     SetStrategy<E>
                      /          \
UnmanagedSetStrategy<E>       ManagedSetStrategy<E>
                                   \
                                SetValueOperator<E> - (BinaryOperator<K> extends SetValueOperator<byte[]>)
                                                      (BooleanOperator<K> extends SetValueOperator<Boolean>)
                                                      (...)
```
Feel free to go through this one at your own pace.

### CODE REVIEW

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmNameTest.java [hunk 1]
```diff
@@ -99,7 +99,7 @@ public void compileModulesWithConflictingPoliciesForNamedClassesFails() {
     // Check that trying to compile two modules with different policies using a mix of `allClasses`
     // and `classes = { ... }` will fail.
     @Test
-    public void compileModulesWithConflictingPoliciesAndMixedClassDefinitionsFails() {
+    public void compileModulesWithConflictingPoliciesAndRealmAnyClassDefinitionsFails() {
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmNameTest.java:102]: This rename is wrong. Mixed did not refer to the mixed datatype here.

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java [hunk 2]
```diff
@@ -189,7 +193,7 @@ public void compileLibraryModulesCustomClasses() {
     }
 
     @Test
-    public void compileAppModuleMixedParametersFail() {
+    public void compileAppModuleRealmAnyParametersFail() {
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmProcessorTest.java:196]: Same, wrong rename

File: realm/realm-annotations-processor/src/test/resources/some/test/NamePolicyMixedClassSettings.java [hunk 3]
```diff
@@ -21,7 +21,7 @@
 import io.realm.annotations.RealmNamingPolicy;
 
 /**
- * Class with mixed settings class/field name settings
+ * Class with realmAny settings class/field name settings
```
Comment (cmelchior) [inline realm/realm-annotations-processor/src/test/resources/some/test/NamePolicyMixedClassSettings.java:24]: Wrong rename

File: realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java [hunk 4]
```diff
@@ -3241,6 +3400,9 @@ private void distinctAllFields(Realm realm, String prefix) {
         for (String fieldName : fieldNames) {
             String field = prefix + fieldName;
             RealmFieldType type = schema.getFieldType(fieldName);
+
+            Log.e("---", "---> type: " + type.toString());
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmQueryTests.java:3404]: Remove?

**Review by cmelchior — APPROVED** [review]
Only very minor comments. Awesome 🚀

File: realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmNameTest.java (hunk 1 above)
Comment (edualonso) [reply realm/realm-annotations-processor/src/test/java/io/realm/processor/RealmNameTest.java:102]: It's amazing you saw these things 🙈

**Pull request merged**